servers/zms/src/main/java/com/yahoo/athenz/zms/NotificationManager.java

```diff
@@ -0,0 +1,172 @@
+        if (notificationServiceFactoryClass != null) {
+            NotificationServiceFactory notificationServiceFactory;
+            try {
+                notificationServiceFactory = (NotificationServiceFactory) Class.forName(notificationServiceFactoryClass).newInstance();
+                notificationService = notificationServiceFactory.create();
+                init();
+            } catch (InstantiationException | IllegalAccessException | ClassNotFoundException e) {
+                LOGGER.error("Invalid NotificationServiceFactory class: " + notificationServiceFactoryClass + " error: " + e.getMessage());
+            }
+        }
+    }
+
+    private void init() {
+        userDomain = System.getProperty(ZMSConsts.ZMS_PROP_USER_DOMAIN, ZMSConsts.USER_DOMAIN);
+        userDomainPrefix = userDomain + ".";
+        if (notificationService != null) {
+            scheduledExecutor = Executors.newScheduledThreadPool(1);
+            scheduledExecutor.scheduleAtFixedRate(new PendingMembershipApprovalReminder(), 0, 1, TimeUnit.DAYS);
+        }
+    }
+
+    NotificationManager(final DBService dbService, final NotificationServiceFactory notificationServiceFactory) {
+        this.dbService = dbService;
+        this.notificationServiceFactory = notificationServiceFactory;
+        notificationService = this.notificationServiceFactory.create();
+        init();
+    }
+
+    void shutdown() {
+        if (scheduledExecutor != null) {
+            scheduledExecutor.shutdownNow();
+        }
+    }
+
+    void generateAndSendPostPutMembershipNotification(String domain, String org, Boolean auditEnabled, Boolean selfserve, Map<String, String> details) {
+        if (isNotificationFeatureAvailable()) {
+            Set<String> recipients = new HashSet<>();
+            if (auditEnabled == Boolean.TRUE) {
+                //get recipient role(s) from audit domain
+                Role domainRole = dbService.getRole(ZMSConsts.SYS_AUTH_AUDIT_DOMAIN, ZMSConsts.AUDIT_APPROVER_ROLE_PREFIX + org + "." + domain, false, true, false);
+                Role orgRole = dbService.getRole(ZMSConsts.SYS_AUTH_AUDIT_DOMAIN, ZMSConsts.AUDIT_APPROVER_ROLE_PREFIX + org, false, true, false);
+                if (domainRole != null) {
+                    recipients.addAll(domainRole.getRoleMembers().stream()
+                            .map(RoleMember::getMemberName).collect(Collectors.toSet()));
+                }
+                if (orgRole != null) {
+                    recipients.addAll(orgRole.getRoleMembers().stream()
+                            .map(RoleMember::getMemberName).collect(Collectors.toSet()));
+                }
+            } else if (selfserve == Boolean.TRUE) {
+                // get admin role from the request domain
+                Role adminRole = dbService.getRole(domain, "admin", false, true, false);
+                recipients.addAll(adminRole.getRoleMembers().stream()
```

Comment:
so for all three cases - lines 91 95 and here we should check that we're notifying users only.

while 91 and 91 - we control so probably there won't be any services, with user domains, not sure what could be there so we should have the checks.